### PR TITLE
Volume controller with tests and some API

### PR DESCRIPTION
**Pull request by ignaczalexander**
Implemented the main Volume methods in Server. Added a few tests, but because some other controllers were not implemented I couldn't test everything I wanted. Will do after merging.

### CODE REVIEW

File: GeorgiaTech/Test/VolumeControllerTests.cs [hunk 1]
```diff
@@ -35,13 +35,13 @@ public void InsertInsertsCorrectly()
             {
                 var insertedVolume = context.Volumes.FirstOrDefault(v => v.MaterialId == materialId && v.CurrentLocationId == currentLocationId && v.HomeLocationId == homeLocationId);
                 Assert.That(insertedVolume, Has
-                    .Property("MaterialId")
+                    .Property(nameof(Volume.MaterialId))
                     .EqualTo(materialId)
                     .And
-                    .Property("HomeLocationId")
+                    .Property(nameof(Volume.HomeLocationId))
```
Comment (Derasm) [inline GeorgiaTech/Test/VolumeControllerTests.cs:43]: nice change - much less magic-strings!

File: GeorgiaTech/Api/Controllers/VolumeController.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+using Api.Models;
+using Server.Controllers;
+// For more information on enabling Web API for empty projects, visit https://go.microsoft.com/fwlink/?LinkID=397860
+
+namespace Api.Controllers
+{
+    [Route("[controller]")]
+    public class VolumeController : Controller
+    {
+        private readonly Server.Controllers.VolumeController volumeController;
```
Comment (Derasm) [inline GeorgiaTech/Api/Controllers/VolumeController.cs:16]: Fairly certain this should be returned through the ControllerFactory, though i can be wrong here.

File: GeorgiaTech/Api/Controllers/VolumeController.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        }
+
+        public Models.Volume BuildVolume(Server.Models.Volume volume)
+        {
+            var modelVolume = new Volume();
+            modelVolume.VolumeId = volume.VolumeId;
+
+            if (volume.HomeLocation != null)
+            {
+                var homeAddress = new Models.Address
```
Comment (Derasm) [inline GeorgiaTech/Api/Controllers/VolumeController.cs:55]: This should be made through the interface for the Address - to ensure you have dependency inversion

File: GeorgiaTech/Api/Controllers/VolumeController.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                    Zip = volume.HomeLocation.ZipCode,
+                    City = volume.HomeLocation.Zip.City
+                };
+                modelVolume.HomeLocation = homeAddress;
+                modelVolume.HomeLocationId = volume.HomeLocationId;
+            }
+       
+            if (volume.CurrentLocation != null)
+            {
+                var currentAddress = new Models.Address
```
Comment (Derasm) [inline GeorgiaTech/Api/Controllers/VolumeController.cs:69]: Same as above: Through interface.

File: GeorgiaTech/Server/Controllers/VolumeController.cs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+{
+    public class VolumeController: IVolumeController
+    {
+        private IMaterialController materialController;
+        private IAddressController addressController;
+        private readonly GTLContext _context;
+
+        public VolumeController(GTLContext context)
+        {
+            materialController = ControllerFactory.CreateMaterialController(context);
```
Comment (Derasm) [inline GeorgiaTech/Server/Controllers/VolumeController.cs:17]: Fucking perfect.

File: GeorgiaTech/Server/Controllers/VolumeController.cs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        /// <returns>The number of rows changed</returns>
+        public int Delete(Volume volume)
+        {
+            using var transaction = _context.Database.BeginTransaction();
+
+            try
+            {
+                _context.Remove(volume);
+                var changedRows = _context.SaveChanges();
+                transaction.Commit();
```
Comment (Derasm) [inline GeorgiaTech/Server/Controllers/VolumeController.cs:62]: Savechanges already implements transactions - which means this is largely unnecessary. Simultaneously, Transactions can't be tested in an InMemoryDatabase

File: GeorgiaTech/Server/Controllers/VolumeController.cs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+                throw;
+
+            }
+        }
+
+        /// <summary>
+        /// Find all the volumes in the database
+        /// </summary>
+        /// <returns>A list of volumes</returns>
+        public List<Volume> FindAll()
```
Comment (Derasm) [inline GeorgiaTech/Server/Controllers/VolumeController.cs:78]: Well done - looks like it was a bitch to figure out.

File: GeorgiaTech/Server/Controllers/VolumeController.cs [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+                return volume;
+            }
+            catch
+            {
+                throw new NullReferenceException(message: $"Volume with id: {ID} not found");
+            }
+    
+        }
+
+        public Volume FindByType(Volume t)
```
Comment (Derasm) [inline GeorgiaTech/Server/Controllers/VolumeController.cs:151]: if this is not going to be implemented ( as others are actually useful) - add the [obsolete] above it.

File: GeorgiaTech/Server/Controllers/VolumeController.cs [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+            _context.Volumes.Add(volume);
+            return _context.SaveChanges();
+        }
+
+        /// <summary>
+        /// Updates a volume entity in the database
+        /// </summary>
+        /// <param name="volume"></param>
+        /// <returns>The number of rows changed</returns>
+        public int Update(Volume volume)
```
Comment (Derasm) [inline GeorgiaTech/Server/Controllers/VolumeController.cs:172]: Same as before: Savechanges already implements transactions, so this is largely unnecessary.

File: GeorgiaTech/Api/Controllers/VolumeController.cs (hunk 3 above)
Comment (ignaczalexander) [reply GeorgiaTech/Api/Controllers/VolumeController.cs:55]: This is not the Entity Address model, but the API one.

**Pull request merged**